ls1intum/Artemis
# Enable new automatic assessment approach for modeling exercises

### Checklist

### Motivation and Context


In this PR we enable Compass again to help tutors with automatic assessments. Additionally, we change the automatic assessment approach in order to provide fairer assessments to the students.

### Description

We enable semi-automatic assessments in Artemis again and change the automatic assessment approach in the way that Compass does not submit complete automatic assessments anymore. Instead it tries to automatically assess as many parts of a model as possible for every modeling submission which then has to be reviewed and completed by a tutor. See the following two activity diagrams illustrating the old and the new approach:

**Old approach**
![Modeling Assessment Workflow (old)](https://user-images.githubusercontent.com/5919138/59554124-d99c1380-8f9e-11e9-830e-4642cab82e5e.png)
**New approach**
![Modeling Assessment Workflow (new)](https://user-images.githubusercontent.com/5919138/59554134-ede01080-8f9e-11e9-9239-a50a8f85556e.png)

**Client-side Changes**
If there is automatic feedback when a tutor starts a new assessment, a message is displayed indicating that parts of the model could already be assessed automatically. The message also makes clear that the tutor has to carefully review the automatic parts as he is solely responsible for the whole assessment. Additionally, all model elements that could already be assessed automatically are highlighted in a blueish color in the model. See the screenshot below for a better understanding of how the assessment editor with automatic assessment looks like.

**Server-side Changes**
The automatic assessment is triggered
* for a new submission right after it is submitted and
* for every submission without manual assessment when a new manual assessment is submitted. 

In the automatic assessment process Compass tries to assign automatic feedback for every model element. In detail, the creation of automatic feedback works as follows:
* Automatic feedback for a model element can only be assigned if Compass already has manual feedback for the corresponding similarity set (see _Knowledge_ in the activity diagram above).
* Automatic feedback is assigned to a model element if the confidence for the corresponding similarity set is at least 80%, i.e. 80% of the existing manual feedback for the similarity set contain the exact same credits. If the confidence is smaller than 80%, no automatic feedback for this element will be added to the automatic assessment.
* The credits of automatic feedback will be chosen according to the confidence, i.e. the credits that 80+% of the manual feedback elements have in common will be chosen. No credits of existing feedback will be merged.
* The feedback text of automatic feedback will be chosen according to the confidence as follows: The longest feedback text from the 80+% of the manual feedback elements of the confidence will be chosen. We assume that longer feedback text means more information/feedback for the students. No feedback text of existing feedback will be merged.

Further, we extended the documentation and refactored/cleaned up the code regarding the automatic assessment in order to increase the quality and comprehensibility of the code..

**Note:** Compass only supports UML class diagrams at the moment. Modeling submissions of any other diagram type have to be assessed completely manually. We will extend Compass to support other diagram types in the near future.


### Steps for Testing

1. Log in to ArTEMiS as student 1
2. Submit model
3. Log in to ArTEMiS as student 2
4. Submit another model containing parts of model submission of student 1
5. Log in to ArTEMiS as a tutor
6. Assess first submission
7. Start assessment for second submission and see automatic assessment for model elements contained in both models
8. Log in to ArTEMiS as student 3
9. Submit another model also containing parts of the first or second submission
10. Log in to ArTEMiS as a tutor
11. Start assessment for new submission and see automatic assessment

### Screenshots


Modeling assessment editor with highlighted automatic feedback
![automatic feedback highlighting](https://user-images.githubusercontent.com/5919138/60216867-83b34f80-986b-11e9-8c4e-d7841b1464ed.png)
